### PR TITLE
Ensure newly created SAC files are valid

### PR DESCRIPTION
**Pull request by jkmacc-LANL**
Bugfix PR for #1204 .

The rewritten `io.sac` doesn't automatically write some important SAC headers, which is a regression, and makes many SAC files created "from scratch" invalid.  This is captured in a new (poorly-named) test `test_valid_sac_from_minimal_existing_sac_header`.

### CODE REVIEW

**Comment by QuLogic** [comment]
Is this finished now? Builds are not passing

**Comment by jkmacc-LANL** [comment]
I've fixed the original issue involving the required headers that weren't being written, but not yet fixed the "merging" of an existing SAC header and the current Trace header.  I was pulled away for a few nights.  Didn't mean to hold anything up.

**Comment by jkmacc-LANL** [comment]
This is passing, but not quite done.

**Comment by jkmacc-LANL** [comment]
I'm satisfied with this PR, barring future suggestions/comments in #1204 .

File: obspy/io/sac/tests/test_core.py [hunk 1]
```diff
@@ -739,6 +740,53 @@ def test_writing_to_file_like_objects_with_obspy(self):
...
+        tr.stats.sac.stla = 1.
+        tr.stats.sac.stlo = 2.
+
+        with NamedTemporaryFile() as tf:
+            tempfile = tf.name
+            tr.write(tempfile, format='SAC')
+            tr1 = read(tempfile)[0]
+
+        # starttime made its way to SAC file
+        # XXX: can't do this. conflicts with XY tests
```
Comment (QuLogic) [inline obspy/io/sac/tests/test_core.py]: Not sure what this comment means.

File: obspy/io/sac/tests/test_core.py [hunk 2]
```diff
@@ -739,6 +740,53 @@ def test_writing_to_file_like_objects_with_obspy(self):
...
+        self.assertEqual(tr1.stats.sac.nzyear, t.year)
+        self.assertEqual(tr1.stats.sac.nzjday, t.julday)
+        self.assertEqual(tr1.stats.sac.nzhour, t.hour)
+        self.assertEqual(tr1.stats.sac.nzmin, t.minute)
+        self.assertEqual(tr1.stats.sac.nzsec, t.second)
+        # don't want to bother with nzmsec b/c of truncation
```
Comment (QuLogic) [inline obspy/io/sac/tests/test_core.py]: You could do an almost-equal check.

File: obspy/io/sac/util.py [hunk 3]
```diff
@@ -268,78 +284,100 @@ def obspy_to_sac_header(stats, keep_sac_header=True):
...
-        and valid reftime are found, the new b and e will be properly
-        referenced to it. If keep_sac_header is False, a new SAC header is
-        constructed from only information found in the stats dictionary, with
-        some other default values introduced.  It will be an iztype 9 ('ib')
-        file, with small adjustments for micro/milliseconds issues.
+    :param keep_sac_header: If keep_sac_header is True, old stats.sac
+        header values are kept, and a minimal set of values are updated from
+        the stats dictionary according to these guidelines:
+        * npts, delta always come from stats
+        * If an old reftime are found and valid, the new b and e will be made
```
Comment (QuLogic) [inline obspy/io/sac/util.py]: s/are/is/

File: obspy/io/sac/util.py [hunk 4]
```diff
@@ -268,78 +284,100 @@ def obspy_to_sac_header(stats, keep_sac_header=True):
...
+            # use stats.starttime as the reftime, and set b, e
+            # ObsPy issue 1204
+            # TODO: consolidate relative-time header list in header.py
+            relhdrs = ['t'+str(i) for i in range(10)] + ['a', 'f']
+            NR = all([header.get(hdr) in (None, HD.SNULL) for hdr in relhdrs])
+            if header.get('iztype') == 9 and NR:
+                reftime = stats['starttime']
+                nztimes, microsecond = utcdatetime_to_sac_nztimes(reftime)
+                header.update(nztimes)
+                header['b'] = (microsecond * 1e-6) if microsecond else 0.0
```
Comment (QuLogic) [inline obspy/io/sac/util.py]: Does this `if` actually do anything? `microsecond` should be a float, no?

File: obspy/io/sac/util.py [hunk 5]
```diff
@@ -268,78 +284,100 @@ def obspy_to_sac_header(stats, keep_sac_header=True):
...
-            # Assume an iztype 9/'ib' type file: move the reftime to the
-            # starttime and assume that the old and new 1st sample times are
-            # the same
+            # TODO: is this needed anymore?
             pass
+
+        # merge some values from stats if they're missing in the SAC header
+        # ObsPy issue 1204
+        if header.get('kstnm') in (None, HD.SNULL):
+            header['kstnm'] = stats['station'] if stats['station']\
```
Comment (QuLogic) [inline obspy/io/sac/util.py]: You can shorten this to `stats['station'] or HD.SNULL`.

File: obspy/io/sac/util.py (hunk 4 above)
Comment (jkmacc-LANL) [inline obspy/io/sac/util.py]: microsecond is an integer number of microseconds.

File: obspy/io/sac/util.py (hunk 5 above)
Comment (jkmacc-LANL) [inline obspy/io/sac/util.py]: That's what I get for coding after midnight :smile:

File: obspy/io/sac/util.py (hunk 4 above)
Comment (QuLogic) [inline obspy/io/sac/util.py]: Right, so `if microsecond` just means `if microsecond != 0` in that case, and `0 * 1e-6 == 0.0` anyway.

Comment (jkmacc-LANL) [inline obspy/io/sac/util.py]: Yes, thank you.

File: obspy/io/sac/tests/test_core.py [hunk 6]
```diff
@@ -739,6 +741,53 @@ def test_writing_to_file_like_objects_with_obspy(self):
...
+        self.assertEqual(tr1.stats.sac.nzyear, nztimes['nzyear'])
+        self.assertEqual(tr1.stats.sac.nzjday, nztimes['nzjday'])
+        self.assertEqual(tr1.stats.sac.nzhour, nztimes['nzhour'])
+        self.assertEqual(tr1.stats.sac.nzmin, nztimes['nzmin'])
+        self.assertEqual(tr1.stats.sac.nzsec, nztimes['nzsec'])
+        self.assertEqual(tr1.stats.sac.nzmsec, nztimes['nzmsec'])
+        self.assertEqual(tr1.stats.sac.kstnm, 'AAA')
+        self.assertEqual(tr1.stats.sac.knetwk, 'XX')
+        self.assertEqual(tr1.stats.sac.kcmpnm, 'BHZ')
+        # self.assertEqual(tr1.stats.khole, '00')
```
Comment (krischer) [inline obspy/io/sac/tests/test_core.py]: This should probably be uncommented.

File: obspy/io/sac/util.py [hunk 7]
```diff
@@ -268,104 +284,114 @@ def obspy_to_sac_header(stats, keep_sac_header=True):
...
-            # assume that the old and new 1st sample times are the same
+            # can't determine reftime or absolute time shift.
             msg = "Old header has invalid reftime."
             warnings.warn(msg)
+
+            # If no relative headers are set and the iztype was 9, we're OK to
+            # use stats.starttime as the reftime, and set b, e
+            # ObsPy issue 1204
+            # TODO: consolidate relative-time header list in header.py
+            relhdrs = ['t'+str(i) for i in range(10)] + ['a', 'f']
```
Comment (krischer) [inline obspy/io/sac/util.py]: Please add spaces around the `+`. I'm not sure why `flake8` does not catch some things.

**Comment by krischer** [comment]
This good to go from my point of view asides my very minor nitpicks. Thanks a lot!

**Comment by jkmacc-LANL** [comment]
Thanks, all!
